misc/metrics/src/protocol_stack.rs

```diff
@@ -0,0 +1,86 @@
+use libp2p_core::multiaddr::{Multiaddr, Protocol};
+use prometheus_client::encoding::text::Encode;
+
+#[derive(Encode, Hash, Clone, Eq, PartialEq)]
+pub struct Label {
+    address_stack: String,
+}
+impl Label {
```

Comment:
Why not implement `From` for `Label`?

``` rust
impl From<&Multiaddr> for Label {
```

Reply:
I think we had this at some point and I recommended to move away from it. Unless we are actually using `From` as an abstraction somewhere, using a generic function has downsides in terms of clarity. It also encourages the use of `.into` at callsites which makes it hard to see, what this is actually being converted into.

Reply:
Ah yes, that sounds familiar. I probably resolved & hid the comment when I committed the recommended change, but it's probably still above somewhere.

Personally I think I prefer the From approach in this case - since the struct has such an incredibly narrow usage I don't think there's too much confusion about what `into` is doing. But of course I'll go with whatever the consensus is.

Reply:
`From` would be nice if we could change the signature of `get_or_create` to accept an `impl Into<S>` so you don't have to call `.into`.

Unfortunately we can't do that without requiring ownership on _every_callsite which would require allocations even though we only need ownership inside the `get_or_create` function for the first call. Meaning a lot of these allocations would be wasted.

I'd be in favor of not obfuscating what we are converting here:

1. Using `.into` makes it hard to find all usages of `Labels`. Currently, I have to delete the `From` impl and see where the compiler fails.
2. I find this:
```rust
                    self.listen_addresses
                        .get_or_create(&protocol_stack::Labels::new(&listen_addr))
                        .inc();
```
much clearer to read than this:
```rust
                    self.listen_addresses
                        .get_or_create(&listen_addr.into())
                        .inc();
```
The latter I can glance over whereas with the former, my brain stops and is like: "Hang on, `into()` what?". And I have to recall the function signature of `get_or_create`, check the type of `listen_addr` and search for `From` implementations to understand what is happening here.

3. If we ever need more data for the labels, `From` would need to be using a tuple at which point we need a constructor anyway.
4. Out of principle, I think we should not be depending on abstractions (`From`), if we don't use them. Traits are only useful in generic code, so unless we can make a function that uses `T: Into` where we can pass a `&Multiaddr`, I think this doesn't qualify.